### PR TITLE
Add support for multi-connection downloads via aria2c

### PR DESCRIPTION
**Pull request by r15ch13**
This change adds support for multi-connection downloads via aria2c. (requested in https://github.com/lukesampson/scoop/issues/2306)
Scoop will use aria2 by default if it's installed locally or globally to speed all downloads.

- [x] Support current cache behaviour
- [x] Disable aria2 via scoops config (`scoop config enable_aria2 false`)
- [x] Added some color to the installation output
- [x] Proxy support (need someone who can test which [proxy settings](https://aria2.github.io/manual/en/html/aria2c.html#http-ftp-sftp-options) are required)
- [x] Change aria2 parameters via scoops config

The output changes from the previous progress bar to simply writing the aria2 output.

aria2-scoop-demo
[still image](https://user-images.githubusercontent.com/432127/41479688-45b3abac-70cc-11e8-8ddb-a26b4995e0a7.png)

Use the following command for constant tests:
`.\bin\scoop.ps1 uninstall godot; .\bin\scoop.ps1 cache rm godot; .\bin\scoop.ps1 install godot`

### CODE REVIEW

File: lib/core.ps1 [hunk 1]
```diff
@@ -92,6 +92,27 @@ function installed_apps($global) {
...
 
+function aria2_path() {
+    # normal path to executable
+    $aria2 = "$(versiondir 'aria2' 'current' $false)\aria2c.exe"
+    if(Test-Path($aria2)) {
+        return $aria2
+    }
+
+    # global path to executable
+    $aria2 = "$(versiondir 'aria2' 'current' $false)\aria2c.exe"
```
Comment (rasa) [inline lib/core.ps1]: This is the same as line 97.  Shouldn’t `$false` be `$true` here?

File: lib/install.ps1 [hunk 2]
```diff
@@ -147,6 +147,116 @@ function do_dl($url, $to, $cookies) {
...
+    $data = @{}
+    $urls = @(url $manifest $architecture)
+
+    # aria2 input file
+    $urlstxt = "$cachedir\$app.txt"
+    $urlstxt_content = ''
+    $has_downloads = $false
+
+    # aria2 options
+    $options = @(
```
Comment (rasa) [inline lib/install.ps1]: Users may want to tweak these defaults. Let’s add a config setting eventually.

File: lib/install.ps1 [hunk 3]
```diff
@@ -147,6 +147,116 @@ function do_dl($url, $to, $cookies) {
...
+        # handle aria2 console output
+        Write-Host "Starting download with aria2 ..."
+        $prefix = "Download: "
+        Invoke-Expression $aria2 | ForEach-Object {
+            if([String]::IsNullOrWhiteSpace($_)) {
+                # skip blank lines
+                return
+            }
+            Write-Host $prefix -NoNewline
+            if($_.EndsWith('download completed.')) {
```
Comment (rasa) [inline lib/install.ps1]: Will fail if aria2’s language isn’t English.

File: lib/install.ps1 (hunk 2 above)
Comment (r15ch13) [reply lib/install.ps1]: Should we use Scoops config settings or use a custom [aria2.conf](https://aria2.github.io/manual/en/html/aria2c.html#aria2-conf)? Should we allow aria2 to load the config file from `HOME` or disable it with [--no-conf](https://aria2.github.io/manual/en/html/aria2c.html#cmdoption-no-conf)?

Comment (rasa) [reply lib/install.ps1]: Good questions. I would say allow aria2 to use its default config file (`$HOME/.aria2/aria2.conf` or `$XDG_CONFIG_HOME/aria2/aria2.conf` if not found, `$XDG_CONFIG_HOME` defaults to `$HOME/.config`, see https://github.com/aria2/aria2/blob/e0a827ff98e665a118e2001dfe29b14847a33070/src/util.cc#L1808 )
In addition, perhaps we have a config setting, say `aria2_options`, to define scoop specific options?

**Comment by jas99** [comment]
Awesome, working for me!!! 👍
How to increase number of threads?

Thx

File: lib/install.ps1 (hunk 2 above)
Comment (jas99) [reply lib/install.ps1]: The desire behaviour must be scoop config taking precedence over aria2.conf. Maybe via script merge aria2.conf with scoop config and create new config file to get this behaviour. Things like number of pieces of downloads should be exposed on root of scoop config.

Thx

**Comment by jas99** [comment]
Error report:
```
Installing 'dbeaver' (5.1.1) [64bit]
Starting download with aria2 ...
Download: [#0e0a4c 0B/0B CN:1 DL:0B]
Download: 06/19 20:36:54 [ERROR] CUID#7 - Download aborted. URI=https://dbeaver.jkiss.org/files/5.1.1/dbeaver-ce-5.1.1-win32.win32.x86_64.zip
Download: Exception: [AbstractCommand.cc:351] errorCode=22 URI=https://dbeaver.io/files/5.1.1/dbeaver-ce-5.1.1-win32.win32.x86_64.zip
Download:   -> [HttpSkipResponseCommand.cc:240] errorCode=22 The response status is not successful. status=400
Download: Download Results:
Download: gid   |stat|avg speed  |path/URI
Download: ======+====+===========+=======================================================
Download: 0e0a4c|ERR |       0B/s|C:/Users/jas/scoop/cache/dbeaver#5.1.1#https_dbeaver.jkiss.org_files_5.1.1_dbeaver-ce-5.1.1-win32.win32.x86_64.zip
Download: Status Legend:
Download: (ERR):error occurred.
Download: aria2 will resume download if the transfer is restarted.
Download: If there are any errors, then see the log file. See '-l' option in help/man page for details.
ERROR Download failed!
DEBUG https://dbeaver.jkiss.org/files/5.1.1/dbeaver-ce-5.1.1-win32.win32.x86_64.zip
    out=dbeaver#5.1.1#https_dbeaver.jkiss.org_files_5.1.1_dbeaver-ce-5.1.1-win32.win32.x86_64.zip

DEBUG C:\Users\jas\scoop\apps\aria2\current\aria2c.exe --input-file='C:\Users\jas\scoop\cache\dbeaver.txt' --dir='C:\Users\jas\scoop\cache' --user-agent='Scoop/1.0 (+http://scoop.sh/) PowerShell/5.1 (Windows NT 10.0; Win64; x64; Desktop)' --allow-overwrite=true --auto-file-renaming=false --retry-wait=2 --split=20 --max-connection-per-server=5 --min-split-size=5M --console-log-level=warn --enable-color=false
```

**Comment by r15ch13** [comment]
@jas99 that is actually a problem with dbeaver. dbeaver.jkiss.org redirects to dbeaver.io and causes a 400 Bad Request with aria2. Changing the URL to dbeaver.io works. (fixed in https://github.com/lukesampson/scoop-extras/commit/b227cd2dc0a5055fc3d5ba765b31e5fbf3f4e547)

**Comment by rasa** [comment]
It sure doesn’t jump out in the error message it’s a 400 error, does it?

**Comment by r15ch13** [comment]
- Disabled `aria2.conf` with `--no-conf=true` to prevent odd behavior if a user has one set up in `$HOME/.aria2/`.
- More clear error messages by translating arias error code (https://aria2.github.io/manual/en/html/aria2c.html#exit-status)
- Added referer and cookie headers
- Made some changes to `get_config()`

**Comment by rasa** [comment]
@r15ch13 Looks great! We may want to eventually add an `aria2-options` to allow users to specify options we haven’t provided them access to. For example:
```
scoop config aria2-options “--all-proxy=<proxy> --connect-timeout=<sec>”
```

**Comment by r15ch13** [comment]
@rasa like so? 😄

**Comment by rasa** [comment]
@r15ch13 Lgtm!

**Comment by yqu212** [comment]
Can a user specify a proxy for each app? For example, 

`scoop install vim --http-proxy="http://proxy"`

Some apps need not proxy. Thus a global proxy is not suitable for  some case.

**Comment by r15ch13** [comment]
@yqu212 this is not even supported by the current installation process. This is something for another feature i guess. 😄

**Comment by yqu212** [comment]
@r15ch13 Ok. I see. By the way, when will this pull request be merged into the master branch?

**Comment by xeijin** [comment]
This is great, I do miss the wget-style `[======>     ]` progress bar though  :( a quick play around with aria2c seems to suggest it doesn't have something similar implemented.

**Comment by xeijin** [comment]
Did this end up going in? Just noticed some of the appveyor tests failed

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DesigningKnights** [comment]
Is this going in soon ?

**Comment by r15ch13** [comment]
Currently, I don't have the time and information I need to implement the Proxy part of this. It would be great If someone could test this with different proxy settings. :smile:

**Comment by se35710** [comment]
@r15ch13 I have tested with nginx (no Connect support, so all tls connections fail), and with squid (with connect support), and it's working great!

```
C:> .\bin\scoop.ps1 config proxy 172.16.1.254:3128
C:> .\bin\scoop.ps1 install eclipse-javascript
Installing 'eclipse-javascript' (4.8.0) [64bit]
Starting download with aria2 ...
Download: [#5cb76c 0B/0B CN:1 DL:0B]
Download: [#5cb76c 704KiB/172MiB(0%) CN:5 DL:792KiB ETA:3m41s]
Download: [#5cb76c 14MiB/172MiB(8%) CN:5 DL:7.4MiB ETA:21s]
Download: [#5cb76c 32MiB/172MiB(18%) CN:5 DL:11MiB ETA:12s]
Download: [#5cb76c 51MiB/172MiB(29%) CN:5 DL:13MiB ETA:9s]
Download: [#5cb76c 69MiB/172MiB(40%) CN:5 DL:14MiB ETA:7s]
Download: [#5cb76c 88MiB/172MiB(51%) CN:5 DL:15MiB ETA:5s]
Download: [#5cb76c 105MiB/172MiB(61%) CN:5 DL:15MiB ETA:4s]
Download: [#5cb76c 122MiB/172MiB(71%) CN:5 DL:15MiB ETA:3s]
Download: [#5cb76c 140MiB/172MiB(81%) CN:5 DL:15MiB ETA:2s]
Download: [#5cb76c 152MiB/172MiB(88%) CN:4 DL:15MiB ETA:1s]
Download: [#5cb76c 157MiB/172MiB(91%) CN:4 DL:15MiB]
Download: [#5cb76c 163MiB/172MiB(94%) CN:3 DL:14MiB]
Download: [#5cb76c 170MiB/172MiB(99%) CN:1 DL:13MiB]
Download: [#5cb76c 171MiB/172MiB(99%) CN:1 DL:11MiB]
Download: Download Results:
Download: gid   |stat|avg speed  |path/URI
Download: ======+====+===========+=======================================================
Download: 5cb76c|OK  |    11MiB/s|C:/Users/mathi/scoop/cache/eclipse-javascript#4.8.0#https_www.eclipse.org_downloads_download.php_file_technology_epp_downloads_release_photon_R_eclipse-javascript-photon-R-win32-x86_64.zip_r_1_dl.7z
Download: Status Legend:
Download: (OK):download completed.
Checking hash of download.php ... ok.
Extracting dl.7z ... done.
Linking ~\scoop\apps\eclipse-javascript\current => ~\scoop\apps\eclipse-javascript\4.8.0
Creating shortcut for Eclipse IDE for JavaScript and Web Developers (eclipse.exe)
'eclipse-javascript' (4.8.0) was installed successfully!

C:> .\bin\scoop.ps1 config proxy 172.16.1.250:3128
C:> .\bin\scoop.ps1 uninstall eclipse-javascript ; .\bin\scoop.ps1 cache rm eclipse-javascript ; .\bin\scoop.ps1 install eclipse-javascript
Installing 'eclipse-javascript' (4.8.0) [64bit]
Starting download with aria2 ...
Download: 08/04 00:13:09 [ERROR] CUID#7 - Download aborted. URI=https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip&r=1#/dl.7z
Download: Exception: [AbstractProxyResponseCommand.cc:75] errorCode=1 Proxy connection failed.
Download: Download Results:
Download: gid   |stat|avg speed  |path/URI
Download: ======+====+===========+=======================================================
Download: 4637ce|ERR |       0B/s|C:/Users/mathi/scoop/cache/eclipse-javascript#4.8.0#https_www.eclipse.org_downloads_download.php_file_technology_epp_downloads_release_photon_R_eclipse-javascript-photon-R-win32-x86_64.zip_r_1_dl.7z
Download: Status Legend:
Download: (ERR):error occurred.
Download: aria2 will resume download if the transfer is restarted.
Download: If there are any errors, then see the log file. See '-l' option in help/man page for details.
ERROR Download failed! (Error 1) An unknown error occurred
DEBUG https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip&r=1#/dl.7z
    referer=https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip&r=1#/
    dir=C:\Users\mathi\scoop\cache
    out=eclipse-javascript#4.8.0#https_www.eclipse.org_downloads_download.php_file_technology_epp_downloads_release_photon_R_eclipse-javascript-photon-R-win32-x86_64.zip_r_1_dl.7z

DEBUG C:\Users\mathi\scoop\apps\aria2\current\aria2c.exe --input-file='C:\Users\mathi\scoop\cache\eclipse-javascript.txt' --user-agent='Scoop/1.0 (+http://scoop.sh/) PowerShell/5.1 (Windows NT 10.0; Win64; x64; Desktop)' --allow-overwrite=true --auto-file-renaming=false --retry-wait=2 --split=5 --max-connection-per-server=5 --min-split-size=5M --console-log-level=warn --enable-color=false --no-conf=true --all-proxy='172.16.1.250:3128'

C:> .\bin\scoop.ps1 install foxe
Installing 'foxe' (2.4.2) [64bit]
Starting download with aria2 ...
Download: [#26f82d 64KiB/403KiB(15%) CN:1 DL:91KiB ETA:3s]
Download: [#26f82d 352KiB/403KiB(87%) CN:1 DL:199KiB]
Download: Download Results:
Download: gid   |stat|avg speed  |path/URI
Download: ======+====+===========+=======================================================
Download: 26f82d|OK  |   219KiB/s|C:/Users/mathi/scoop/cache/foxe#2.4.2#http_www.firstobject.com_foxe242.zip
Download: Status Legend:
Download: (OK):download completed.
Checking hash of foxe242.zip ... ok.
Extracting foxe242.zip ... done.
Linking ~\scoop\apps\foxe\current => ~\scoop\apps\foxe\2.4.2
Creating shim for 'foxe'.
'foxe' (2.4.2) was installed successfully!
```

squid logs:
```
1533334064.145   4761 172.16.1.16 TCP_TUNNEL/200 4679 CONNECT www.eclipse.org:443 - HIER_DIRECT/198.41.30.198 -
1533334064.447   3908 172.16.1.16 TCP_TUNNEL/200 517 CONNECT www.eclipse.org:443 - HIER_DIRECT/198.41.30.198 -
1533334064.480   3940 172.16.1.16 TCP_TUNNEL/200 4224 CONNECT www.eclipse.org:443 - HIER_DIRECT/198.41.30.198 -
1533334064.486   3945 172.16.1.16 TCP_TUNNEL/200 4221 CONNECT www.eclipse.org:443 - HIER_DIRECT/198.41.30.198 -
1533334065.603   1969 172.16.1.16 TCP_MISS/206 21283895 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334067.356   1748 172.16.1.16 TCP_MISS/206 16777721 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334068.664   1304 172.16.1.16 TCP_MISS/206 15729143 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334069.323    655 172.16.1.16 TCP_MISS/206 7340534 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334069.844   8727 172.16.1.16 TCP_MISS_ABORTED/206 28315953 GET http://mirror.switch.ch/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/130.59.113.36 application/zip
1533334069.907    578 172.16.1.16 TCP_MISS/206 6291958 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334071.982  11673 172.16.1.16 TCP_MISS_ABORTED/200 22066802 GET http://mirror.switch.ch/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/130.59.113.36 application/zip
1533334072.206  10763 172.16.1.16 TCP_MISS_ABORTED/206 7294448 GET http://mirror.switch.ch/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/130.59.113.36 application/zip
1533334072.839    632 172.16.1.16 TCP_MISS/206 4273174 GET http://mirror.dkm.cz/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/86.49.49.49 application/zip
1533334073.071   3212 172.16.1.16 TCP_MISS/206 6291889 GET http://mirror.switch.ch/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/130.59.113.36 application/zip
1533334074.956  13532 172.16.1.16 TCP_MISS_ABORTED/206 16781618 GET http://mirror.switch.ch/eclipse/technology/epp/downloads/release/photon/R/eclipse-javascript-photon-R-win32-x86_64.zip - HIER_DIRECT/130.59.113.36 application/zip
```

nginx logs:
```
172.16.1.16 - - [04/Aug/2018:00:13:09 +0200] "CONNECT www.eclipse.org:443 HTTP/1.1" 400 182 "-" "-"
172.16.1.16 - - [04/Aug/2018:00:13:09 +0200] "CONNECT www.eclipse.org:443 HTTP/1.1" 400 182 "-" "-"
172.16.1.16 - - [04/Aug/2018:00:13:09 +0200] "CONNECT www.eclipse.org:443 HTTP/1.1" 400 182 "-" "-"
172.16.1.16 - - [04/Aug/2018:00:13:09 +0200] "CONNECT www.eclipse.org:443 HTTP/1.1" 400 182 "-" "-"
172.16.1.16 - - [04/Aug/2018:00:13:09 +0200] "CONNECT www.eclipse.org:443 HTTP/1.1" 400 182 "-" "-"
172.16.1.16 - - [04/Aug/2018:00:16:09 +0200] "GET http://www.firstobject.com/foxe242.zip HTTP/1.1" 200 412759 "http://www.firstobject.com/" "Scoop/1.0 (+http://scoop.sh/) PowerShell/5.1 (Windows NT 10.0; Win64; x64; Desktop)"
```

**Comment by se35710** [comment]
sourceforge is broken though, both with and without proxy defined:

```
C:> .\bin\scoop.ps1 install console2
Installing 'console2' (2.0.0.148) [64bit]
Starting download with aria2 ...
Download: [#81c45a 0B/0B CN:1 DL:0B]
Download: Download Results:
Download: gid   |stat|avg speed  |path/URI
Download: ======+====+===========+=======================================================
Download: 81c45a|OK  |   400KiB/s|C:/Users/mathi/scoop/cache/console2#2.0.0.148#https_downloads.sourceforge.net_project_console_console-devel_2.00_Console-2.00b148-Beta_64bit.zip
Download: Status Legend:
Download: (OK):download completed.
Checking hash of Console-2.00b148-Beta_64bit.zip ... ERROR Hash check failed!
App:       extras/console2
URL:       https://downloads.sourceforge.net/project/console/console-devel/2.00/Console-2.00b148-Beta_64bit.zip
Expected:  675c3c82695a97dafe164960a50dfbe6c13d0a900b382efb7090090138d22933
Actual:    ce421ad8395690a3df8901fdbef785537bbfc94793dfa25160719d650884e529
```
Manually downloading it with aria2 works, and gives the correct checksum. I have tested with multiple sourceforge projects, they all get checksum errors with aria2 when running `scoop install`.

**Comment by r15ch13** [comment]
@se35710 can you provide your proxy settings for nginx/squid so I can test this?

**Pull request merged**